File: docs/validation/cpp/app/src/scenarios/s06_builder.cpp
// S-06: Build synthetic config with MockConfigBuilder and verify fields
#include "s06_builder.hpp"
#include "machine_config/machine_config.hpp"

#include <cmath>

namespace s06 {

using namespace machine_config;

scenarios::Result run(const std::filesystem::path&, const std::filesystem::path&) {
    try {
        MockConfigBuilder builder;
        builder.laser_count = 2;
        MachineConfig cfg = builder.build();

        if (cfg.optical_trains.size() != 2) {
            return {false, "optical_trains count: " + std::to_string(cfg.optical_trains.size())};
        }

        const auto& r0 = cfg.optical_trains[0].scanner.scan_head_rotation;
        const auto& r1 = cfg.optical_trains[1].scanner.scan_head_rotation;
        if (!r0 || std::abs(*r0) > 0.001) return {false, "train[0].scan_head_rotation mismatch"};
        if (!r1 || std::abs(*r1 - 180.0) > 0.001) return {false, "train[1].scan_head_rotation mismatch"};
        if (cfg.meta.machine_name.empty()) return {false, "machine_name is empty"};

        const auto& clearbox = cfg.optical_trains[0].optional_components.clearbox;
        if (!clearbox) return {false, "clearbox is absent"};
        if (!clearbox->correction_data) return {false, "correction_data is absent"};

        GridCell center = (*clearbox->correction_data)[128][128][0];
        if (!center || !std::isfinite(*center) || std::abs(*center - 2.0) > 0.01) {
            return {false, "correction_data center: expected ~2.0"};
        }

        auto tmp = scenarios::makeTempPath(".h5");
        MachineConfigWriter{cfg}.write(tmp);
        MachineConfig rb = MachineConfigReader{tmp}.parse();

        if (rb.optical_trains.size() != 2) {
            return {false, "readback trains: " + std::to_string(rb.optical_trains.size())};
        }
        if (rb.meta.machine_name != cfg.meta.machine_name) {
            return {false, "machine_name changed after roundtrip"};
        }

        return {true, "2-laser build OK, center~=2.0, roundtrip OK"};
    } catch (const std::exception& e) {
        return {false, std::string("exception: ") + e.what()};
    }
}

} // namespace s06

File: docs/validation/go/app/scenarios/s06_builder.go
package scenarios

// S-06: Build synthetic config with MockConfigBuilder and verify fields
//
// ID:       S-06
// Action:   Build a 2-laser config -> verify fields -> save to temp -> re-read.
// Expected: 2 trains, rotations 0/180, machine_name non-empty,
//           correction_data centre cell ~2.0 (Gaussian peak), roundtrip OK.

import (
	"fmt"
	"math"
	"os"

	mc "machine-config-go"
)

func RunS06Builder(_, _ string) (bool, string) {
	cfg := mc.NewMockConfigBuilder().Build()

	if len(cfg.OpticalTrains) != 2 {
		return false, fmt.Sprintf("optical_trains count: %d", len(cfg.OpticalTrains))
	}

	r0 := cfg.OpticalTrains[0].Scanner.ScanHeadRotation
	if r0 == nil || math.Abs(*r0) > 0.001 {
		return false, fmt.Sprintf("train[0].scan_head_rotation: %s", fmtF64Ptr(r0))
	}
	r1 := cfg.OpticalTrains[1].Scanner.ScanHeadRotation
	if r1 == nil || math.Abs(*r1-180.0) > 0.001 {
		return false, fmt.Sprintf("train[1].scan_head_rotation: %s", fmtF64Ptr(r1))
	}
	if cfg.Meta.MachineName == "" {
		return false, "machine_name is empty"
	}

	cb := cfg.OpticalTrains[0].OptionalComponents.Clearbox
	if cb == nil || cb.CorrectionData == nil {
		return false, "clearbox or correction_data is nil"
	}
	grid := *cb.CorrectionData
	center := grid[128][128][0]
	if center == nil || math.Abs(*center-2.0) > 0.01 {
		return false, fmt.Sprintf("correction_data center: expected ~2.0, got %s", fmtF64Ptr(center))
	}

	tmpPath, err := tempH5("mcl_go_s06")
	if err != nil {
		return false, fmt.Sprintf("tempfile: %v", err)
	}
	defer os.Remove(tmpPath)

	if err := mc.NewWriter().Write(cfg, tmpPath); err != nil {
		return false, fmt.Sprintf("write failed: %v", err)
	}
	rb, err := mc.NewReader(tmpPath).Parse()
	if err != nil {
		return false, fmt.Sprintf("readback failed: %v", err)
	}
	if len(rb.OpticalTrains) != 2 {
		return false, fmt.Sprintf("readback trains: %d", len(rb.OpticalTrains))
	}
	if rb.Meta.MachineName != cfg.Meta.MachineName {
		return false, "machine_name changed after roundtrip"
	}

	return true, "2-laser build OK, center~2.0, roundtrip OK"
}
